dynamiq/storages/vector/pgvector/pgvector.py

```diff
@@ -0,0 +1,648 @@
+
+        Returns:
+            list[Document]: List of retrieved Document objects.
+
+        Raises:
+            ValueError: If query_embedding is empty or filter format is incorrect.
+        """
+        if not query_embedding:
+            msg = "query_embedding must be a non-empty list"
+            raise ValueError(msg)
+
+        if len(query_embedding) != self.dimension:
+            msg = f"query_embedding must be of dimension {self.dimension}"
+            raise ValueError(msg)
+
+        vector_function = self.vector_function
+        content_key = content_key or self.content_key
+        embedding_key = embedding_key or self.embedding_key
+
+        if vector_function not in PGVectorVectorFunction:
+            msg = f"Invalid vector function: {vector_function}"
+            raise ValueError(msg)
+
+        query_embedding = self._convert_query_embedding_to_pgvector_format(query_embedding)
+
+        # Generate the score calculation based on the vector function
+        score_definition = VECTOR_FUNCTION_TO_SCORE_DEFINITION[vector_function].format(
+            embedding_key=embedding_key, query_embedding=query_embedding
+        )
+        score_definition = f"{score_definition} AS score"
+
+        # Do not select the embeddings if exclude_document_embeddings is True
+        select_fields = f"id, {content_key}, metadata" if exclude_document_embeddings else "*"
+
+        # Build the base SELECT query with score
+        base_select = SQL("SELECT {fields}, {score} FROM {schema_name}.{table_name}").format(
+            fields=SQL(select_fields),
+            score=SQL(score_definition),
+            schema_name=Identifier(self.schema_name),
+            table_name=Identifier(self.table_name),
+        )
+
+        # Handle filters if they exist
+        where_clause = SQL("")
+        params = ()
+        if filters:
+            where_clause, params = _convert_filters_to_query(filters)
+
+        # Determine sort order based on vector function type
+        is_distance_metric = vector_function in ["l2_distance", "l1_distance"]
+
+        # Sort by score in ascending order if using a distance metric
+        # as the smaller the distance, the more similar the vectors are
+        sort_order = "ASC" if is_distance_metric else "DESC"
+
+        # Build the ORDER BY and LIMIT clause
+        order_by = SQL(" ORDER BY score {sort_order} LIMIT {limit}").format(
+            sort_order=SQL(sort_order), limit=SQLLiteral(top_k)
+        )
+
+        # Combine all parts into final query
+        sql_query = base_select + where_clause + order_by
+
+        with self._get_connection() as conn:
+            with conn.cursor(row_factory=dict_row) as cur:
+                result = self._execute_sql_query(sql_query, params, cursor=cur)
+                records = result.fetchall()
+
+                documents = self._convert_query_result_to_documents(records)
+                return documents
+
+    def __del__(self):
```

Comment:
Please add a separate `def close` method for this purposes. We are trying to close all connection in ConnectionManeger by calling this method if it present:
https://github.com/dynamiq-ai/dynamiq/blob/9134b277128f2f59b2b9884ae726a787e8afaa1e/dynamiq/connections/managers.py#L134-L142

Reply:
Does it mean that the `__del__()` method is redundant here and I don't even have to add a separate `close()` method to close the connection as it will be closed automatically by the `ConnectionManager`?

Reply:
let's add like this:
```
    def close(self):
        if hasattr(self, "_conn") and self._conn is not None and not self._conn.closed:
            self._conn.close()

    def __del__(self):
        """Close the connection when the object is deleted."""
        self.close()
```

Reply:
Thanks, it's done now.